File: 2. Strings/3. Valid Palindrome And Almost Palindrome/almost-palindrome.py
def check_palindrome(s: str, start: int, end: int) -> bool:
    while start <= end:
        if s[start] != s[end]:
            return False
        start += 1
        end -= 1
    return True


def validPalindrome(s: str) -> bool:
    start, end = 0, len(s) - 1
    while start <= end:
        if s[start] != s[end]:
            return check_palindrome(s, start + 1, end) or check_palindrome(
                s, start, end - 1
            )
        start += 1
        end -= 1
    return True

File: 2. Strings/3. Valid Palindrome And Almost Palindrome/almost-palindrome.cpp
#include <bits/stdc++.h>
#include <iostream>

using namespace std;

class Solution {
public:
  bool checkPalindrome(string s, int start, int end) {
    while (start <= end) {
      if (s[start] != s[end]) {
        return false;
      }
      start++;
      end--;
    }
    return true;
  }

  bool validPalindrome(string s) {
    int start = 0;
    int end = s.length() - 1;

    while (start <= end) {
      if (s[start] != s[end]) {
        return checkPalindrome(s, start + 1, end) or
               checkPalindrome(s, start, end - 1);
      }
      start++;
      end--;
    }
    return true;
  }
};

int main() { return 0; }